### ISSUE TITLE
Contention in SerializingExecutor$TaskRunner.run

### ISSUE THREAD

**Issue by carl-mastrangelo** [grpc/grpc-java#2122]
Not sure why there is much contention here, but there does appear to be some.  This is very tiny compared to the rest of the other contention points.

```
      flat  flat%   sum%        cum   cum%
     0.21s 0.0054% 97.67%   2853.93s 74.89%  io.grpc.internal.SerializingExecutor$TaskRunner.run SerializingExecutor.java:142

```

**Comment by pgrosu** [comment]
Are you concerned that it's spending 0.21 seconds inside the `run` function?  You can have up to 4 `Runnable` in your `waitQueue`, end the `internalLock` monitor object is being synchronized which is an semi-significant overhead as it's part of the JVM.  So that can happen up to 4 times + 1 when it gets to `finally`.  There are other ways to perform concurrency without using `synchronized` if necessary, but you are saying that you have to support Java 6 so you might have to write the concurrency mechanism from scratch.

**Comment by carl-mastrangelo** [comment]
.21 seconds spent waiting on the lock to become free.  That is latency added to RPCs, likely tail latency.

**Comment by pgrosu** [comment]
Here's a little program to demo this, but I'm not getting anywhere as high as 210 milliseconds, but it does increase with the number of threads.

```Java
class MonitoredLock {
   public void print(){
    try {
        System.out.println( "...Lock is Synchronized..." );
     } catch (Exception e) {
         System.out.println(" Thread is interrupted. ");
     }
   }
}

class SynchronizedThread extends Thread {

   private Thread thread;
   private String threadName;
   int semaphore = 10;
   MonitoredLock  ML;

   SynchronizedThread( String name,  MonitoredLock refML){
       threadName = name;
       ML = refML;
   }

   public void run() {
     while ( semaphore > 0 ) {
       long tick = System.currentTimeMillis();
       synchronized (ML) {
          ML.print();
          long elapsed = System.currentTimeMillis() - tick;
          System.out.println("Thread " +  threadName + " -> elapsed during sync: " + elapsed );
       }
       long elapsed = System.currentTimeMillis() - tick;
       System.out.println("Thread " +  threadName + " -> elapsed outside sync: " + elapsed );
       semaphore = semaphore - 1;
      }
   }

   public void start ()
   {
      System.out.println("Starting " +  threadName );
      if (thread == null)
      {
         thread = new Thread (this, threadName);
         thread.start ();
      }
   }

}

public class SynchronizedTest {
   public static void main(String args[]) {

      MonitoredLock ML = new MonitoredLock();

      SynchronizedThread T1 = new SynchronizedThread( " 1 ", ML );
      SynchronizedThread T2 = new SynchronizedThread( " 2 ", ML );
      SynchronizedThread T3 = new SynchronizedThread( " 3 ", ML );
      SynchronizedThread T4 = new SynchronizedThread( " 4 ", ML );

      T1.start();
      T2.start();
      T3.start();
      T4.start();

      try {
         T1.join();
         T2.join();
         T3.join();
         T4.join();
      } catch( Exception e) {
         System.out.println("Interrupted");
      }
   }
}
```
Below is the result of running it:

```
> javac SynchronizedTest.java && java SynchronizedTest
Starting  1
Starting  2
...Lock is Synchronized...
Starting  3
Thread  1  -> elapsed during sync: 0
...Lock is Synchronized...
Starting  4
Thread  3  -> elapsed during sync: 0
Thread  1  -> elapsed outside sync: 1
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 0
Thread  4  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 3
Thread  4  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 1
Thread  3  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 2
Thread  1  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 3
Thread  3  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 0
Thread  4  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 2
Thread  2  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 3
Thread  4  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 1
Thread  3  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 2
Thread  1  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 3
Thread  3  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 1
Thread  4  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 2
Thread  2  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 3
Thread  4  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 1
Thread  3  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  1  -> elapsed during sync: 3
Thread  1  -> elapsed outside sync: 3
Thread  3  -> elapsed outside sync: 3
...Lock is Synchronized...
Thread  4  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 3
Thread  2  -> elapsed during sync: 4
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 5
Thread  4  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 2
Thread  3  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 4
Thread  1  -> elapsed during sync: 4
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 5
Thread  3  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 1
Thread  4  -> elapsed during sync: 2
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 3
Thread  2  -> elapsed during sync: 4
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 4
Thread  4  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  4  -> elapsed outside sync: 1
Thread  3  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 2
Thread  1  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 4
Thread  3  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  3  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 3
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 3
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 2
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 0
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
Thread  2  -> elapsed outside sync: 1
> 
```

By running only two threads, naturally the elapsed times are smaller:

```
> java SynchronizedTest
Starting  1
Starting  2
...Lock is Synchronized...
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 2
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 2
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 0
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 0
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 0
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  2  -> elapsed outside sync: 1
Thread  1  -> elapsed during sync: 1
...Lock is Synchronized...
Thread  1  -> elapsed outside sync: 1
Thread  2  -> elapsed during sync: 0
Thread  2  -> elapsed outside sync: 1
>
```

**Comment by ejona86** [comment]
@pgrosu, the .21s was _cumulative_ across a ~5 minute run on a 32 core machine (each time the lock was not available, the amount of time the thread had to wait for the lock was added to a total). So it isn't that big of problem, but it was detectable. #2120 is 5000x worse :-).

**Comment by carl-mastrangelo** [comment]
@pgrosu Thank you for looking into this.  I  created this issue so I wouldn't forget about it later.  It isn't urgent or important.

**Comment by pgrosu** [comment]
@ejona86 Thank you for the clarification, but where do you find a 5000 core machine and how do you run JVM on it?  I'm trying to go through the flow of the threads using the code.

@carl-mastrangelo It's still an interesting problem.  The thing about #2120 looks to be  a different method than here:
```
io.grpc.internal.ManagedChannelImpl$3.getTransport ManagedChannelImpl.java:598
```
versus
```
io.grpc.internal.SerializingExecutor$TaskRunner.run SerializingExecutor.java:142
```
I'm assuming a threadpool is used.  Would a larger threadpool be more practical?  It'd imagine that it would be hard to change the code to become more event-driven through an infinite loop scanning loop with a queue of events that become ready.  Then threads for non-concurrent I/O channels can be used unless multiplexing is needed.

**Comment by pgrosu** [comment]
Doing a quick back-of-the envelope calculation in R, it looks like each core has a `0.19` contribution to the overall-time bottleneck.  Are you guys running a Kubernetes-type implementation simulating a 5000 core setup, where the bandwidth throughput and CPU overhead is contributing to the slowdown?  Maybe some things can be cached so that the minimum required communication happens.

```R
> elapsed.time <- c( 0.21, 932.07 )
> num.of.cores <- c( 32, 5000 )
> benchmark.data <- data.frame( elapsed.time, num.of.cores )
> lm( elapsed.time ~ num.of.cores, data=benchmark.data )

Call:
lm(formula = elapsed.time ~ num.of.cores, data = benchmark.data)

Coefficients:
 (Intercept)  num.of.cores  
     -5.7923        0.1876  

> 
```

**Comment by ejona86** [comment]
I've got a lockless implementation in the works. It works, but I'm trying to simplify the error cases, and it will need tests.

**Comment by pgrosu** [comment]
Eric, this is fantastic news - you just made my morning!

**Comment by ejona86** [comment]
This was resolved in #2858.
